[Mission 3] 3단계 미션 구현 - 김희영/1팀

heyoung0218 opened the pull request:
<!-- PR 제목 예시:
[missionX] X단계 미션 구현 - 000/0팀
-->

## 구현 사항
-  Nav-bar 검색 기능 구현 
- 메인 페이지 인기 영화 슬라이드 외 추가로 5개 장르 영화 추가
- 페이지 반응형 디자인 추가
- 다크모드/라이트 모드 토글식으로 구현

## 어려웠던 점
- css 인라인 스타일을 tailwind로 바꿔서 하다가 에러가 발생함. (해결 못함 tailwind 4.1.10 error 에러 관해서 아는거 있으면 조언 부탁드립니다.)
- 다크모드 적용시 상세 페이지는 비교적 쉽게 해결 했지만, 메인 페이지 적용이 너무 안돼서 애먹었습니다.
 ( 로그인, 회원가입 페이지는 못만들었습니다.)

## 구현 이미지
- 
![스크린샷 2025-06-22 23 29 38](https://github.com/user-attachments/assets/5833cbdd-5cf4-401b-b18c-41e5b683f940)
![스크린샷 2025-06-22 23 29 48](https://github.com/user-attachments/assets/51224fb4-504f-4e8e-a5aa-f769df2e3bd4)
![스크린샷 2025-06-22 23 30 02](https://github.com/user-attachments/assets/6e3cd4d6-6cde-4745-b71c-6d5ecc647498)
![스크린샷 2025-06-22 23 30 14](https://github.com/user-attachments/assets/f145f72d-e997-4f78-a3de-4fcd8ee0f97f)

heyoung0218 committed:
feat: API 호출 방식 변경 - API 키에서 읽기 액세스 토큰 방식으로 수정 및 헤더 적용

heyoung0218 committed:
feat: 검색어 입력시 API호출 제어하기 위한 useDebounce 커스텀 훅 구현으로 검색 입력 지연 처리

heyoung0218 committed:
feat: TMDb 검색 API 함수 fetchSearchMovies 추가

heyoung0218 committed:
feat: App 컴포넌트에서 검색어 기반 영화 데이터 로드 기능 추가

heyoung0218 committed:
fix: MovieApi.js , useDebounce훅 변수명 오타 수정

heyoung0218 committed:
feat: Navbar에 검색 입력창 및 onSearch 이벤트 핸들링 추가

heyoung0218 committed:
feat: Layout 컴포넌트에서 검색어 상태 관리 및 전달

heyoung0218 committed:
feat: App에서 useOutletContext로 NavBar 검색어 받아 사용

heyoung0218 committed:
fix: 공백 검색 방지 보완
- App.jsx debouncedSearch.trim()으로 공백 검색 방지 추가
- movieApi.js fetchSearchMovies 공백만 입력된 경우 빈 배열 반환 처리 추가

heyoung0218 committed:
refactor: useDebounce 변수명 개선

heyoung0218 committed:
feat: fetchMoviesByGenre API 함수 추가 (장르별 영화 조회)

heyoung0218 committed:
feat: GenreMovieList 컴포넌트 추가 (장르별 영화 목록 출력)

heyoung0218 committed:
feat: 메인 페이지에 즐겨찾는 장르별 영화 카드 목록 출력 기능 추가

heyoung0218 committed:
refactor: App.jsx내부 로직을 useAppLogic 훅으로 분리하여 코드 가독성 및 재사용성 향상

heyoung0218 committed:
feat: 다크모드/라이트모드 상태 관리용 ThemeContext 생성

heyoung0218 committed:
feat: ThemeProvider로 전역 다크모드 설정 추가

heyoung0218 committed:
chore: Tailwind CSS 적용을 위한 postcss.config.js 추가

heyoung0218 committed:
refactor: Tailwind 기반으로 index.css 리팩토링 및 불필요한 스타일 제거

heyoung0218 committed:
fix: PostCSS오류 발생하여  설정을 CJS 형식으로 변경하여 Tailwind 호환성 해결

heyoung0218 committed:
feat: 다크모드/라이트모드 상태 관리용 ThemeContext 생성

heyoung0218 committed:
feat: ThemeProvider로 전역 다크모드 설정 추가

heyoung0218 committed:
"feat: NavBar에 다크모드/라이트모드 토글 버튼 추가"

"refactor: NavBar 컴포넌트 Tailwind 기반으로 전면 리팩토링 및 디자인 개선"

- 다크모드 토글 스타일 개선 및 아이콘 추가
- 반응형 레이아웃 구성
- 잘못된 useTheme import 경로 수정

heyoung0218 committed:
style: App.jsx Tailwind CSS 적용 및 다크모드 스타일 추가

heyoung0218 committed:
style: Tailwind 적용 오류로 임시 비활성화 및 기본 CSS로 전환

- postcss.config.cjs에서 Tailwind 플러그인 제거
- 추후 Tailwind 오류 해결 후 다시 활성화 예정

heyoung0218 committed:
refactor: Tailwind CSS 제거 및 전반적인 UI를 기본 CSS 스타일로 전환

- App.jsx: Tailwind 클래스 제거, 배경/텍스트 색상 및 레이아웃을 인라인 스타일로 변경
- MovieCard.jsx: 영화 카드 UI를 기본 스타일로 재구성, 별점 표시 유지
- NavBar.jsx: 반응형 Tailwind 레이아웃 제거, 검색창/버튼/다크모드 토글에 기본 스타일 적용
- SkeletonMovieCard.jsx: 스켈레톤 로딩 UI에 Tailwind 제거, 애니메이션 및 색상 직접 지정
- MovieDetail.jsx: 상세 정보 페이지에서 Tailwind 완전히 제거, 구조 유지하며 클래식 스타일 적용

heyoung0218 committed:
refactor: App.jsx 리팩토링으로 인기 영화 카드 크기 확대 및 슬라이드 화살표 복원
- 인기 영화 카드에 large prop 적용
- Swiper 슬라이드 네비게이션 스타일 복구
- 다크모드 배경/텍스트 전환 및 부드러운 애니메이션 추가
- 미디어 쿼리로 반응형 대응 (768px, 1024px 기준)

heyoung0218 committed:
feat: MovieCard에 large prop 추가하여 인기 영화 카드 크기 조정
- 슬라이드용 large 카드 크기 및 폰트 확대
- 장르 리스트는 기존 사이즈 유지

heyoung0218 committed:
fix: 슬라이드 카드 간격 및 반응형 문제 해결

- spaceBetween으로 카드 간격 조정
- 고정 width로 카드 크기 통일
- breakpoints로 slidesPerView 반응형 적용
- navigation 옵션 복구해 화살표 표시

heyoung0218 committed:
refactor: ThemeContext에 toggleTheme 함수 추가 및 다크모드 제어 로직 개선

heyoung0218 committed:
feat: NavBar에 로고 추가 및 반응형 정렬 적용, 다크모드 toggleTheme 적용

heyoung0218 committed:
style: Layout.jsx에 다크모드 배경 및 텍스트 색상 인라인 스타일 적용

Kwonyeojiny commented on src/hooks/useDebounce.js:
`useDebounce` ✅

Kwonyeojiny commented on src/index.css:
```diff
@@ -1,72 +1,27 @@
-  -webkit-font-smoothing: antialiased;
-  -moz-osx-font-smoothing: grayscale;
-}
+/* tailwind 적용 오류
+/* @tailwind base;
+@tailwind components;
+@tailwind utilities; */
```
해당 코드는 v3에서 넣어주는 코드이고 
tailwind v4에서는 index.css에 
```
@import 'tailwindcss';
```
요거 한줄 넣어주시면 됩니다!

Kwonyeojiny commented on src/pages/App.jsx:
```diff
@@ -9,67 +9,134 @@ import "swiper/css/pagination";
-    }
-    loadMovies();
-  }, []);
+  const { searchTerm, darkMode } = useOutletContext();
+  const debouncedSearch = useDebounce(searchTerm, 500);
```
현재 useSearchParams를 사용하지 않으시고 검색 input의 value를 useState를 이용해서 업데이트 해주시고 useOutletContext로 가져와서 디바운스 체크하고 api를 불러오는 형식으로 하시면서 검색어 하나를 위해 너무 복잡한 구조가 된 것 같습니다
이렇게 되면 유지보수도 어렵고 코드가 얽혀 의존관계가 복잡해집니다

검색 기능은 
input의 값을 useDebounce를 활용하여 일정시간 대기한 후, 해당 값(`debouncedSearch`)을 URL 쿼리 파라미터에 반영하고,
페이지에서는 useSearchParams를 활용해 쿼리 값을 읽어 검색 API를 호출하면
현재 희영님께서 작성하신 것보다 간결하고 직관적인 검색 기능을 구현할 수 있습니다.

리팩토링..... 시키면 쓰러지시겠죠..?!

Kwonyeojiny commented:
검색되는 이미지도 추가해주실 수 있나요?!

heyoung0218 committed:
feat: Layout.jsx에서 다크모드 상태 및 토글 관리, NavBar 및 하위 라우트에 상태 및 검색어 컨텍스트 전달 추가
fix: Layout.jsx SSR 환경에서 초기 스타일 플래시 방지를 위해 mounted 상태로 렌더링 지연 처리

heyoung0218 committed:
feat: MovieCard에 large prop 추가하여 카드 크기 조정 가능하도록 개선, 다크모드 개선

heyoung0218 committed:
refactor: ThemeContext에 toggleTheme 함수 추가 및 다크모드 제어 로직 개선

heyoung0218 committed:
feat: 장르별 영화 리스트 10개로 제한 및 성인영화 필터,다크모드/스타일 최적화

heyoung0218 committed:
feat: SkeletonMovieCard에 large prop 추가 및 인라인 다크모드,애니메이션 적용

heyoung0218 committed:
style: MovieDetail.jsx 다크모드 스타일 적용, 로딩 및 에러 처리 UI 인라인 스타일로 개선
fix: MovieDetail.jsx 다크모드 인라인 스타일 적용으로 다크모드 정상 작동하도록 수정

heyoung0218 committed:
fix: 슬라이드 카드 간격 및 반응형 문제 해결

- spaceBetween으로 카드 간격 조정
- 고정 width로 카드 크기 통일
- breakpoints로 slidesPerView 반응형 적용
- navigation 옵션 복구해 화살표 표시
- 다크모드 미적용 문제 해결

heyoung0218 committed:
refactor: 검색어 상태를 useSearchParams 기반으로 변경

- 검색 입력값을 URL 쿼리 파라미터에 반영하도록 NavBar 수정
- App 컴포넌트에서 useSearchParams로 쿼리값 가져와 디바운싱 처리
- Outlet context와 useState 기반의 검색 로직 제거

heyoung0218 committed:
refactor: 검색 기능을 URL 쿼리 기반으로 리팩토링 및 디바운스 적용

    - Navbar에서 input 상태를 useState로 관리하고, useDebounce 적용
    - 디바운스된 값이 있을 경우 URL 쿼리(query)로 반영
    - App.jsx에서 useSearchParams로 query 값을 읽어 API 요청
    - 입력과 쿼리 상태를 분리하여 입력 깨짐 문제 해결

heyoung0218 committed:
feat: 검색 결과 일반 목록 표시 및 검색 결과 없을 시 메시지 추가
- 검색 결과 별도 SearchResultList.jsx 컴포넌트 분리
- 기존 슬라이드 형태 표시를 리스트 형태로 표시
= 검색 결과 없을 때 표시

fix: genreMovies 및 movies 데이터 undefined 참조 방지 처리
fix: NavBar 컴포넌트 import 경로 수정

heyoung0218 committed:
refactor: 불필요한 useEffect로직 제거

heyoung0218 committed:
feat: 상세 페이지에서 검색 기능 추가

- 상세페이지로 이동 시 query 파라미터 유지 
- 상세 페이지(MovieDetail)에서 검색 쿼리 관련 로직 제거
- 상세페이지에서 검색어 표시 유지  

feat: 검색 자동완성 기능 추가 및 뒤로가기/앞으로가기 브라우저 내비게이션 반응 개선

- 검색어 입력 시 추천 영화 자동완성 리스트 표시 (제목 포함 + 인기순 정렬)
- 엔터 입력 시만 검색 navigate 발생하도록 개선 (불필요한 히스토리 최소화)
- 추천 리스트에 투명도(opacity) 적용 및 리스트 점 제거
- 외부 클릭 시 자동완성 리스트 닫히도록 처리
- 브라우저 뒤로가기 및 앞으로가기 기능 정상 작동 보장  

heyoung0218 commented:
검색 기능 리팩토링 했습니다. (검색어 상태를 useSearchParms 기반으로 변경 + 검색되는 이미지 추가) 
리팩토리중 발생한 문제 -> 
메인페이지에서는 검색창이 작동을 하는데 상세페이지에서는 작동을 안함 = 수정함
메인 페이지 검색어 입력시 자동으로 페이지가 이동됨 = 엔터 입력시 입력하게끔 수정함
검색창 자동완성된 영화리스트 클릭시 상세페이지로 이동이 안됨 = 수정함
페이지이동시 뒤로가기/앞으로가기가 제대로 작동하지 않음 = 수정함

Kwonyeojiny commented on src/api/movieApi.js:
```diff
@@ -20,12 +44,42 @@ export async function fetchPopularMovies() {
-      }
-      return await response.json();
+      `${BASE_URL}/movie/${id}?language=ko-KR`, { 
+        method: "GET",
+        headers: getAuthHeaders()
```
api 불러올 때마다 공통되는 해당 옵션을 분리해서 공통 유틸 함수로 관리해주면 코드 재사용성과 유지보수 측면에서 더 좋습니다!
```
{
  method: "GET",
  headers: getAuthHeaders()
}
```
해당 코드를
```
export const  getFetchOptions = () {
  return {
    method: "GET",
    headers: getAuthHeaders(),
  };
}
```
이런 식으로 분리해주면 fetch해줄 때 
```
`${BASE_URL}/movie/${id}?language=ko-KR`, getFetchOptions()
```
이렇게 작성돼서 더 가독성에도 좋죠?!

Kwonyeojiny commented on src/components/Navbar.jsx:
```diff
@@ -1,74 +1,255 @@
-const NavBar = ({ onSearch }) => {
-  const [searchTerm, setSearchTerm] = useState("");
+const NavBar = () => {
+  const [searchParams, setSearchParams] = useSearchParams();
```
searchParams 사용!! 리팩토링🤩

Kwonyeojiny commented on src/hooks/useDebounce.js:
useDebounce 훅도 잘 구현되었네요! 👍🏻

Kwonyeojiny commented:
리팩토링 하시느라 수고많으셨습니다!!

heyoung0218 committed:
fix:  자동완성 클릭 시 빈 화면 문제 해결및 MovieDetail 에서 불필요한 검색어 상태 제거 
 - 자동완성 클릭 후 navigate는 되지만 빈 화면이 나오는 문제 발생 
 - 상세 페이지는 URL의 id만으로 영화 정보 로딩
 - useDebounce, query 상태 제거  

fix: 상세페이지 라우팅 경로 및 네비게이션 경로 일치시키기
 - main.jsx 경로를 details/:id' → 'movie/:id'로 수정 
 - App.jsx 경로를 'details/:id' → 'movie/:id'로 수정하여 자동완성 클릭 시 빈 화면 문제 해결 

heyoung0218 committed:
fix: MovieCard.jsx fallback 클릭 핸들러 추가
 - MovieCard 컴포넌트에서 onClick이 전달되지 않는 경우 대비하여 fallback 클릭 핸들러 추가
